components/script/dom/htmlmediaelement.rs

```diff
@@ -1350,6 +1393,53 @@ impl HTMLMediaElementMethods for HTMLMediaElement {
         self.paused.get()
     }
 
+    /// https://html.spec.whatwg.org/multipage/#dom-media-defaultplaybackrate
+    fn GetDefaultPlaybackRate(&self) -> Fallible<Finite<f64>> {
+        Ok(Finite::wrap(self.defaultPlaybackRate.get()))
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-defaultplaybackrate
+    fn SetDefaultPlaybackRate(&self, value: Finite<f64>) -> ErrorResult {
+        let min_allowed = -64.0;
+        let max_allowed = 64.0;
+        if *value < min_allowed || *value > max_allowed {
+            return Err(Error::NotSupported);
+        }
+
+        if *value != self.defaultPlaybackRate.get() {
+            self.defaultPlaybackRate.set(*value);
+            self.queue_ratechange_event();
+        }
+
+        Ok(())
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-playbackrate
+    fn GetPlaybackRate(&self) -> Fallible<Finite<f64>> {
+        Ok(Finite::wrap(self.playbackRate.get()))
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#dom-media-playbackrate
+    fn SetPlaybackRate(&self, value: Finite<f64>) -> ErrorResult {
+        let min_allowed = -64.0;
+        let max_allowed = 64.0;
+        if *value < min_allowed || *value > max_allowed {
+            return Err(Error::NotSupported);
+        }
+
+        if *value != self.playbackRate.get() {
+            self.playbackRate.set(*value);
+            self.queue_ratechange_event();
+            if self.is_potentially_playing() {
+                if let Err(e) = self.player.set_rate(*value) {
```

Comment:
One last change! Let's set the property and queue the `ratechange` event only if setting the player rate succeeds.

Reply:
With this change in place, it looks like the tests timeouts are back.

Reply:
Ok. I think this may be a misinterpretation (or actually a lack of clarity) of the spec.

So the spec says [here](https://html.spec.whatwg.org/multipage/media.html#dom-media-playbackrate):

> Set playbackRate to the new value, and if the element is potentially playing, change the playback speed.

Changing the playback speed (`self.player.set_rate()`) depends on the element being potentially playing. 

But then it says as well:

> When the defaultPlaybackRate or playbackRate attributes change value (either by being set by script or by being changed directly by the user agent, e.g. in response to user control) the user agent must queue a task to fire an event named ratechange at the media element.

It does not say anything about firing the `ratechange` event only if the element is potentially playing. So we will need to change the attribute value (`self.playbackRate.set(*value)`) and fire the event (`self.queue_ratechange_event()`) **even if the element is not potentially playing**.

The problem with allowing the `playbackRate` attribute change when the media element is not playing is that we will be missing the actual rate change when the media starts playing. So to fix that issue we will need to apply the current `playbackRate` attribute value to the player (`self.player.set_rate(self.playbackRage.get())`) right before the media starts playing.

Reply:
Thank you for the explanations! I opened an issue here: [https://github.com/whatwg/html/issues/4256](https://github.com/whatwg/html/issues/4256)

Reply:
It seems that updating the playback rate of the media element was implicitly mentioned in another paragraph of the API: https://github.com/whatwg/html/issues/4256#issuecomment-453116074

Reply:
Ah, I see! Thanks for following up on that. The code fully reflects what the spec says then. Thanks!